### ISSUE TITLE
Malformed HTTP Header causes exception in twisted.web.http server

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8101]
|<img alt="Javantea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Javantea reported|
|-|-|
|Trac ID|trac#8101|
|Type|defect|
|Created|2015-11-12 03:16:49Z|
|Branch|https://github.com/twisted/twisted/tree/http-server-invalid-header-8101-2|

When a malformed HTTP header with no colon is sent, the twisted http server will throw and catch an exception returning no data to the sender.

[twisted/web/http.py#L427](twisted/web/http.py#L427)

Stack Trace:
```
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 101, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/log.py", line 84, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 611, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 209, in doRead
    return self._dataReceived(data)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 215, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/usr/lib64/python2.7/site-packages/twisted/protocols/basic.py", line 571, in dataReceived
    why = self.lineReceived(line)
  File "/usr/lib64/python2.7/site-packages/twisted/web/http.py", line 1687, in lineReceived
    self.headerReceived(self.__header)
  File "/usr/lib64/python2.7/site-packages/twisted/web/http.py", line 1705, in headerReceived
    header, data = line.split(b':', 1)
exceptions.ValueError: need more than 1 value to unpack
```

repro:
```python
from twisted.web import server
from twisted.internet import reactor
reactor.listenTCP(4009, server.Site(None))
reactor.run()
```
```#!sh
echo -n $'GET / HTTP/1.1\r\nHost: localhost\r\nAAAA\r\n\r\n' | nc localhost 4009
```

Twisted should either send a 400 (like apache), treat it like a header with no value (nginx) or like it wasn't sent (Python).

Attachments:

* [8101patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc5/cc5685aa9d6180ebc03bebe9c0d04dd5bfa21df6/fcc7c97ef5ea5efbea03850491acd2847cdb846a.diff) (2121 bytes) - added by pawelmhm on 2015-11-15 12:14:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8101 8101
type__defect defect
reporter__Javantea Javantea
priority__normal normal
milestone__None None
branch__branches_http_server_invalid_header_8101_2 branches/http-server-invalid-header-8101-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1447298209785249 1447298209785249
changetime__1459752246684149 1459752246684149
version__None None
owner__pawelmhm pawelmhm

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46235]) Branching to http-server-invalid-header-8101.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @pawelmhm|
|-|-|

Many thanks for your patch.

It looks good and I am happy with the fix.

Please note that the twisted/web/topfiles/8101.bugfix file needs a text describing the bugfix.
The text will be part of the final release notes and will help people know about your fix.

I have tried to update the description.


Please check that my text is ok.


If all is ok and test pass, I will merge it.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_4"></a>|@pawelmhm commented|
|-|-|

> I have tried to update the description. Please check that my text is ok.

I'm totally ok with you updating description, but I dont see it anywhere in ticket, did you uploaded new diff somewhere? Should I upload diff with updated bugfix file?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

You can see my changes here https://github.com/twisted/twisted/compare/trunk...http-server-invalid-header-8101 last commit.

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_6"></a>|@pawelmhm commented|
|-|-|

thanks adiroiban! this looks good. 

+1 from me

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46310]) Merge http-server-invalid-header-8101: Fix exception in twisted.web.http.HTTPChannel.headerReceived when malformed data is received.

Author: pawelmhm
Reviewer: adiroiban
Fixes: [#8101](https://github.com/twisted/twisted/issues/8101)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set status to `reopened`|
|-|-|

merged in the wrong branch

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [47134]) Branching to http-server-invalid-header-8101-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set status to `closed`|
|-|-|

(In [47143]) Merge http-server-invalid-header-8101-2: Fix exception in twisted.web.http.HTTPChannel.headerReceived when malformed data is received.

Author: pawelmhm
Reviewer: adiroiban
Fixes: [#8101](https://github.com/twisted/twisted/issues/8101)
